libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/search.hpp

```diff
@@ -198,24 +686,360 @@ namespace hpx { namespace parallel { inline namespace v1 {
     ///           returned. If no subsequence is found, \a last is returned.
     ///
     template <typename ExPolicy, typename Rng1, typename Rng2,
-        typename Pred = detail::equal_to,
-        typename Proj1 = util::projection_identity,
-        typename Proj2 = util::projection_identity,
+        typename Pred = hpx::ranges::equal_to,
+        typename Proj1 = hpx::parallel::util::projection_identity,
+        typename Proj2 = hpx::parallel::util::projection_identity>
+    typename hpx::parallel::util::detail::algorithm_result<ExPolicy,
+        typename hpx::traits::range_iterator<Rng1>::type>::type
+    search_n(ExPolicy&& policy, Rng1&& rng1, std::size_t count, Rng2&& rng2,
+        Pred&& op = Pred(), Proj1&& proj1 = Proj1(), Proj2&& proj2 = Proj2());
+
+}}    // namespace hpx::ranges
+
+#else
+
+namespace hpx { namespace parallel { inline namespace v1 {
+
+    // clang-format off
+    template <typename ExPolicy, typename Rng1, typename Rng2,
+        typename Pred = hpx::ranges::equal_to,
+        typename Proj1 = parallel::util::projection_identity,
+        typename Proj2 = parallel::util::projection_identity,
         HPX_CONCEPT_REQUIRES_(
-            hpx::is_execution_policy<ExPolicy>::value&& hpx::traits::is_range<
-                Rng1>::value&& traits::is_projected_range<Proj1, Rng1>::value&&
-                hpx::traits::is_range<Rng2>::value&& traits::is_projected_range<
-                    Proj2, Rng2>::value&& traits::is_indirect_callable<ExPolicy,
-                    Pred, traits::projected_range<Proj1, Rng1>,
-                    traits::projected_range<Proj2, Rng2>>::value)>
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_range<Rng1>::value &&
+            traits::is_projected_range<Proj1, Rng1>::value &&
+            hpx::traits::is_range<Rng2>::value &&
+            traits::is_projected_range<Proj2, Rng2>::value &&
+            traits::is_indirect_callable<ExPolicy,
+                Pred, traits::projected_range<Proj1, Rng1>,
+                traits::projected_range<Proj2, Rng2>
+            >::value
+        )>
+    HPX_DEPRECATED_V(1, 6, "Please use hpx::ranges::search instead.")
+    // clang-format on
     typename util::detail::algorithm_result<ExPolicy,
         typename hpx::traits::range_iterator<Rng1>::type>::type
-    search_n(ExPolicy&& policy, Rng1&& rng1, std::size_t count, Rng2&& rng2,
-        Pred&& op = Pred(), Proj1&& proj1 = Proj1(), Proj2&& proj2 = Proj2())
+        search(ExPolicy&& policy, Rng1&& rng1, Rng2&& rng2, Pred&& op = Pred(),
+            Proj1&& proj1 = Proj1(), Proj2&& proj2 = Proj2())
+    {
+        return search(std::forward<ExPolicy>(policy), hpx::util::begin(rng1),
```

Comment:
Let's call the algorithm implementation directly here instead of going through the CPO again. Same elsewhere.

Reply:
I didn't change this since it is the deprecated def. Should I change it?

Reply:
Good point! That's up to you. The others could skip the CPO though (if they don't already).